### PR TITLE
feat: pass more options to dagster-aws Resources

### PR DESCRIPTION
**Pull request by jacob-white-simplisafe**
## Summary & Motivation

I am using dagster with localstack for dev testing, and I am unable to pass `endpoint_url` variable to the `SSMResource` as with the `S3Resource`. This PR will allow for more configuration options to be passed to resources that support it.

- `CloudwatchLogsHandler`
- `ECRPublicClient`
- `SecretsManagerResource`
- `SSMResource`

## How I Tested These Changes

The unit tests for the library passed. As the resources that are being changed, all call `boto3.session.Session.client()` verified that the options are in the [documentation for boto3](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/session.html#boto3.session.Session.client).

### CODE REVIEW

**Comment by yuhan** [comment]
@mlarose could you take a look?

File: python_modules/libraries/dagster-aws/dagster_aws/ssm/parameters.py [hunk 1]
```diff
@@ -7,16 +7,36 @@
...
-    max_attempts: int, region_name: Optional[str] = None, profile_name: Optional[str] = None
+    max_attempts: int,
+    region_name: Optional[str] = None,
+    profile_name: Optional[str] = None,
+    endpoint_url: Optional[str] = None,
+    use_ssl: bool = True,
+    aws_access_key_id: Optional[str] = None,
+    aws_secret_access_key: Optional[str] = None,
+    aws_session_token: Optional[str] = None,
+    verify=None,
```
Comment (mlarose) [inline python_modules/libraries/dagster-aws/dagster_aws/ssm/parameters.py:18]: verify can be a boolean or a string according to boto3 documentation.

while verify=False for production use would not be quite advisable, I can envision that someone might want to point to a development endpoint with a self-signed certificate.

do you think it's worth supporting here?

File: python_modules/libraries/dagster-aws/dagster_aws/utils/__init__.py [hunk 2]
```diff
@@ -40,6 +40,29 @@ class ResourceWithBoto3Configuration(ConfigurableResource):
...
+    use_ssl: bool = Field(
+        default=True, description="Whether or not to use SSL. By default, SSL is used."
+    )
+    endpoint_url: Optional[str] = Field(
+        default=None, description="Specifies a custom endpoint for the Boto3 session."
+    )
+    verify: Optional[str] = Field(
+        default=None,
+        description=(
+            "Whether or not to verify SSL certificates. By default SSL certificates are verified."
```
Comment (mlarose) [inline python_modules/libraries/dagster-aws/dagster_aws/utils/__init__.py:52]: A description is slightly misleading vs how to use this. See other comment

**Review by mlarose — CHANGES_REQUESTED** [review]
Thanks for your contribution. I think the verify parameter needs revisiting, but I am open to being convinced otherwise.

File: python_modules/libraries/dagster-aws/dagster_aws/utils/__init__.py (hunk 2 above)
Comment (jacob-white-simplisafe) [reply python_modules/libraries/dagster-aws/dagster_aws/utils/__init__.py:52]: This is the description that I copied from the existing S3Resource
https://github.com/dagster-io/dagster/blob/master/python_modules/libraries/dagster-aws/dagster_aws/s3/resources.py#L36-L43

File: python_modules/libraries/dagster-aws/dagster_aws/ssm/parameters.py (hunk 1 above)
Comment (jacob-white-simplisafe) [reply python_modules/libraries/dagster-aws/dagster_aws/ssm/parameters.py]: As all we are doing is passing it through to boto3, I think it is worth supporting. Certainly not recommended, but don't think it is worth blocking users on.

**Comment by jacob-white-simplisafe** [comment]
It looks like it might be too much work to support the verify parameter. Pydanitic is throwing an error of
```
E   dagster._core.errors.DagsterInvalidPythonicConfigDefinitionError: 
E   Error defining Dagster config class <class 'dagster_aws.ssm.resources.SSMResource'> on field 'verify'.
E   Unable to resolve config type typing.Union[bool, str, NoneType] to a supported Dagster config type.
E   
E   
E   This config type can be a:
E       - Python primitive type
E           - int, float, bool, str, list
E       - A Python Dict or List type containing other valid types
E       - Custom data classes extending dagster.Config
E       - A Pydantic discriminated union type (https://docs.pydantic.dev/usage/types/#discriminated-unions-aka-tagged-unions)
```

I don't have enough experience with pydantic and how dagster handles it to fix the union type. I'm going to revert to only support the bool option type for verify.

**Comment by mlarose** [comment]
i am ok with this, can be improved on later, you're already adding options that were missing

**Comment by jacob-white-simplisafe** [comment]
Getting weird errors locally when running tests for files I haven't touched.
```
==================================================================================================== short test summary info ====================================================================================================
FAILED python_modules/libraries/dagster-aws/dagster_aws_tests/emr_tests/test_pyspark.py::test_local - dagster._core.errors.DagsterResourceFunctionError: Error executing resource_fn on ResourceDefinition pyspark
FAILED python_modules/libraries/dagster-aws/dagster_aws_tests/emr_tests/test_pyspark.py::test_pyspark_emr - dagster._core.errors.DagsterResourceFunctionError: Error executing resource_fn on ResourceDefinition pyspark
FAILED python_modules/libraries/dagster-aws/dagster_aws_tests/pipes_tests/test_pipes.py::test_lambda_s3_pipes - KeyError: 'meta'
```

**Comment by mlarose** [comment]
> Now only use bool values, but getting weird errors locally when running tests for files I haven't touched.
> 
> ```
> ==================================================================================================== short test summary info ====================================================================================================
> FAILED python_modules/libraries/dagster-aws/dagster_aws_tests/emr_tests/test_pyspark.py::test_local - dagster._core.errors.DagsterResourceFunctionError: Error executing resource_fn on ResourceDefinition pyspark
> FAILED python_modules/libraries/dagster-aws/dagster_aws_tests/emr_tests/test_pyspark.py::test_pyspark_emr - dagster._core.errors.DagsterResourceFunctionError: Error executing resource_fn on ResourceDefinition pyspark
> FAILED python_modules/libraries/dagster-aws/dagster_aws_tests/pipes_tests/test_pipes.py::test_lambda_s3_pipes - KeyError: 'meta'
> ```

Buildkite seems happy except for the code linting complaining about unused import.

https://buildkite.com/dagster/unit-tests/builds/677#018f5a02-8f4e-4257-9192-4486ea5c9b93/229-323

> [2024-05-08T21:06:27Z] python_modules/libraries/dagster-aws/dagster_aws/s3/resources.py:1:44: F401 [*] `typing.Union` imported but unused


For the test, I'm inclined to think it's either a dependency you must provide to support the test or something missing in your dev setup, and a false negative. I will trust buildkite's result.

**Review by mlarose — CHANGES_REQUESTED** [review]
- [ ] unused import

**Review by mlarose — APPROVED** [review]

File: python_modules/libraries/dagster-aws/dagster_aws/cloudwatch/loggers.py [hunk 3]
```diff
@@ -34,14 +35,22 @@ def __init__(
...
         log_stream_name,
         aws_region=None,
-        aws_secret_access_key=None,
-        aws_access_key_id=None,
+        aws_access_key_id: Optional[str] = None,
+        aws_secret_access_key: Optional[str] = None,
+        endpoint_url: Optional[str] = None,
+        use_ssl: bool = True,
+        aws_session_token: Optional[str] = None,
+        verify: Optional[bool] =None,
```
Comment (mlarose) [inline python_modules/libraries/dagster-aws/dagster_aws/cloudwatch/loggers.py:43]:
```suggestion
        verify: Optional[bool] = None,
```

File: python_modules/libraries/dagster-aws/dagster_aws/s3/utils.py [hunk 4]
```diff
@@ -23,24 +25,24 @@ def __call__(self, bytes_amount):
...
-    profile_name=None,
-    use_ssl=True,
-    verify=None,
-    aws_access_key_id=None,
-    aws_secret_access_key=None,
-    aws_session_token=None,
+    max_attempts: int,
+    region_name: Optional[str] = None,
+    endpoint_url: Optional[str] = None,
+    use_unsigned_session: bool=False,
```
Comment (mlarose) [inline python_modules/libraries/dagster-aws/dagster_aws/s3/utils.py:31]:
```suggestion
    use_unsigned_session: bool = False,
```

File: python_modules/libraries/dagster-aws/dagster_aws/secretsmanager/secrets.py [hunk 5]
```diff
@@ -7,16 +7,36 @@
...
-    max_attempts: int, region_name: Optional[str] = None, profile_name: Optional[str] = None
+    max_attempts: int,
+    region_name: Optional[str] = None,
+    profile_name: Optional[str] = None,
+    endpoint_url: Optional[str] = None,
+    use_ssl: bool = True,
+    aws_access_key_id: Optional[str] = None,
+    aws_secret_access_key: Optional[str] = None,
+    aws_session_token: Optional[str] = None,
+    verify: Optional[bool]=None,
```
Comment (mlarose) [inline python_modules/libraries/dagster-aws/dagster_aws/secretsmanager/secrets.py:18]:
```suggestion
    verify: Optional[bool] = None,
```

**Review by mlarose — CHANGES_REQUESTED** [review]
oops, i should have checked that ruff format 
sorry, it feels like I am nagging you.

**Review by mlarose — APPROVED** [review]

**Comment by mlarose** [comment]
thanks @jacob-white-simplisafe

**Pull request merged**